internal/conf/jwk.go

```diff
@@ -0,0 +1,192 @@
+	// key's "x", "y" fields cannot be validated unless the "crv" field is populated first.
+	for _, key := range *j {
+		if err := key.PrivateKey.Validate(); err != nil {
+			return err
+		}
+		// symmetric keys don't have public keys
+		if key.PublicKey != nil {
+			if err := key.PublicKey.Validate(); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func GetSigningJwk(config *JWTConfiguration) (*jwk.Key, error) {
+	if config.Keys == nil {
+		symmetricJwk, err := getSymmetricKey(config)
+		if err != nil {
+			return nil, err
+		}
+		return symmetricJwk, nil
+	}
+	for _, key := range config.Keys {
+		if key.PrivateKey.KeyUsage() == "enc" {
+			return &key.PrivateKey, nil
+		}
+	}
+	// return symmetric secret anyway to preserve existing behavior
+	symmetricJwk, err := getSymmetricKey(config)
+	if err != nil {
+		return nil, err
+	}
+	return symmetricJwk, nil
+}
+
+func GetSigningKey(k *jwk.Key) (any, error) {
+	var key any
+	switch (*k).KeyType() {
+	case jwa.OctetSeq:
+		key = []byte{}
+	case jwa.EC:
+		key = ecdh.PrivateKey{}
+	case jwa.RSA:
+		key = rsa.PrivateKey{}
+	case jwa.OKP:
+		// OKP is used for EdDSA keys
+		key = ed25519.PrivateKey{}
+	case jwa.InvalidKeyType:
+		return nil, jwt.ErrInvalidKeyType
+	}
+	if err := (*k).Raw(&key); err != nil {
+		return nil, err
+	}
+	return key, nil
+}
+
+func GetSigningAlg(k *jwk.Key) jwt.SigningMethod {
+	if k == nil {
+		return jwt.SigningMethodHS256
+	}
+
+	switch (*k).Algorithm().String() {
+	case "RS256":
+		return jwt.SigningMethodRS256
+	case "RS512":
+		return jwt.SigningMethodRS512
+	case "ES256":
+		return jwt.SigningMethodES256
+	case "ES512":
+		return jwt.SigningMethodES512
+	case "EdDSA":
+		return jwt.SigningMethodEdDSA
+	}
+
+	// return HS256 to preserve existing behaviour
+	return jwt.SigningMethodHS256
+}
+
+func GetKeyByKid(kid string, config *JWTConfiguration) (any, error) {
+	if kid == config.KeyID {
+		return []byte(config.Secret), nil
+	}
+	if k, ok := config.Keys[kid]; ok {
+		var key interface{}
+		if err := k.PublicKey.Raw(&key); err != nil {
+			return nil, err
+		}
+		return key, nil
+	}
+	return nil, fmt.Errorf("invalid kid: %s", kid)
+}
+
+func getSymmetricKey(config *JWTConfiguration) (*jwk.Key, error) {
+	if config.Secret != "" {
+		bytes, err := base64.StdEncoding.DecodeString(config.Secret)
+		if err != nil {
+			bytes = []byte(config.Secret)
+		}
+		privKey, err := jwk.FromRaw(bytes)
+		if err != nil {
+			return nil, err
+		}
+		if config.KeyID != "" {
+			if err := privKey.Set(jwk.KeyIDKey, config.KeyID); err != nil {
+				return nil, err
+			}
+		}
+		if privKey.Algorithm().String() == "" {
+			if err := privKey.Set(jwk.AlgorithmKey, "HS256"); err != nil {
+				return nil, err
+			}
+		}
+		if privKey.KeyUsage() == "" {
+			if err := privKey.Set(jwk.KeyUsageKey, "enc"); err != nil {
+				return nil, err
+			}
+		}
+		return &privKey, nil
+	}
+	return nil, fmt.Errorf("missing symmetric key")
```

Comment:
You can probably memoize this calculation in a private field in the config struct.

Reply:
Not sure why you're setting the usage to encryption also.

Reply:
ah yeah the intention is to fallback and use the `GOTRUE_JWT_SECRET` - i realised i should just do it in `ApplyDefaults` if `GOTRUE_JWT_KEYS` is nil and just create the symmetric JWK there